include/pgduckdb/pgduckdb_process_latch.hpp

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include <mutex>
+#include <condition_variable>
+
+extern "C" {
+struct Latch;
+extern struct Latch *MyLatch;
+extern int	WaitLatch(Latch *latch, int wakeEvents, long timeout,
+					  uint32_t wait_event_info);
+extern void ResetLatch(Latch *latch);
+
+/* Defined in storage/latch.h */
+#define WL_LATCH_SET (1 << 0)
+#define WL_EXIT_ON_PM_DEATH	 (1 << 5)
+
+/* Defined in utils/wait_event.h */
+#define PG_WAIT_EXTENSION 0x07000000U
+
+}
+
+namespace pgduckdb {
+
+/*
+ * GlobalProcessLatch is used to wait on process postgres latch. First thread that enters will wait for latch while
+ * others will wait until this latch is released.
+ */
+
+struct GlobalProcessLatch {
+public:
+	static void WaitGlobalLatch() {
+		static std::condition_variable cv;
+		static std::mutex lock;
+		if (lock.try_lock()) {
+			WaitLatch(MyLatch, WL_LATCH_SET | WL_EXIT_ON_PM_DEATH, 0, PG_WAIT_EXTENSION);
+			ResetLatch(MyLatch);
+			cv.notify_all();
+			lock.unlock();
+		} else {
+			std::unique_lock<std::mutex> lk(lock);
+			cv.wait(lk);
+		}
```

Comment:
Could you explain a bit what the goal is of this logic. I had expected it to basically be:

1. Take global process lock
2. Call WaitLatch
3. Call ResetLatch
4. Unlock global process lock 

Is this specific logic, for performance reasons? Or is it necessary to avoid concurrency bugs? e.g. is this meant to make sure that if two threads wait for the latch, both get notified instead of one thread "eating" the notification that the other was waiting for?

Reply:
We can wait only on single Latch in process. So first thread arriving will wait on latch while others will wait for condition variable (that will be released once latch waiting is done). Idea was to "simulate" different threads to wait on latch.

Reply:
As discussed over a call, we need some counter here to make sure that one thread doesn't consume the notifcations for another thread. Something like this:
```c
		if (lock.lock()) {
			static global_counter = 0;
			if (previous_counter != global_counter) { /* previous_counter should be thread_local */
				WaitLatch(MyLatch, WL_LATCH_SET | WL_EXIT_ON_PM_DEATH, 0, PG_WAIT_EXTENSION);
				ResetLatch(MyLatch);
				global_counter++;
			}
			previous_counter = global_counter; 
			lock.unlock();
		}
```